### DISCUSSION TITLE
Running async code from a ServiceConfig function

### DISCUSSION THREAD

**Discussion by Cakeday** [actix/actix-web#3181] [Q&A]
Is there a way to call asynchronous code from a service config function?

It almost seems like a 1-to-1 use case compared to invoking the app, except the main function is async.

> Run external configuration as part of the scope building process.
> 
> This function is useful for moving parts of configuration to a different module or library. For example, some of the resource's configuration could be moved to different module.

```rust
use actix_web::{web, middleware, App, HttpResponse};

// this function could be located in different module
fn config(cfg: &mut web::ServiceConfig) {
    cfg.service(web::resource("/test")
        .route(web::get().to(|| HttpResponse::Ok()))
        .route(web::head().to(|| HttpResponse::MethodNotAllowed()))
    );
}

let app = App::new()
    .wrap(middleware::Logger::default())
    .service(
        web::scope("/api")
            .configure(config)
    )
    .route("/index.html", web::get().to(|| HttpResponse::Ok()));
```

I would love to do something like this:

```rust
use actix_web::{web, middleware, App, HttpResponse};

// this function could be located in different module
async fn config(cfg: &mut web::ServiceConfig) {

    let some_async_value = some_async_function().await;

    cfg.service(web::resource("/test")
        .route(web::get().to(|| HttpResponse::Ok()))
        .route(web::head().to(|| HttpResponse::MethodNotAllowed()))
    );
}

let app = App::new()
    .wrap(middleware::Logger::default())
    .service(
        web::scope("/api")
            .configure(config)
    )
    .route("/index.html", web::get().to(|| HttpResponse::Ok()));
```

**Comment by Samielakkad** [comment]
`ServiceConfig::configure` itself is synchronous. It is meant to build the routing tree, not await I/O. So you cannot pass an `async fn config(cfg: &mut ServiceConfig)` directly to `.configure(...)`.

The usual pattern is to do the async work before constructing the `App`, then move/clone the result into a normal config function or closure.

For example, if the async value is global app state:

```rust
#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let some_async_value = some_async_function().await;
    let some_async_value = web::Data::new(some_async_value);

    HttpServer::new(move || {
        App::new()
            .app_data(some_async_value.clone())
            .service(
                web::scope("/api")
                    .configure(config)
            )
    })
    .bind(("127.0.0.1", 8080))?
    .run()
    .await
}

fn config(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::resource("/test")
            .route(web::get().to(test_handler))
    );
}
```

If the async value affects which routes you register, compute it first and capture it in a closure:

```rust
let enabled = load_feature_flags().await;

HttpServer::new(move || {
    let enabled = enabled.clone();

    App::new().service(
        web::scope("/api").configure(move |cfg| {
            if enabled.test_route {
                cfg.service(
                    web::resource("/test")
                        .route(web::get().to(test_handler))
                );
            }
        })
    )
})
```

One important detail: `HttpServer::new` builds one `App` per worker. If you put the async initialization inside the factory, it would need to happen per worker, but the factory is not async. If the resource should be shared, initialize it before `HttpServer::new` and clone a handle/pool/Arc into each worker.

For expensive per-worker setup that truly must happen separately for each worker, use synchronous construction in the factory or store a lazily initialized async resource behind app state and initialize it on first use. But for most config/database/client setup, doing it before `HttpServer::new` is simpler and more predictable.
